sdk/src/utils/font_thumbnail.rs

```diff
@@ -0,0 +1,410 @@
+}
+
+/// Load font data into the font database, returning the ID of the last font
+fn load_font_data(font_db: &mut Database, data: Vec<u8>) -> Result<ID> {
+    font_db.load_font_data(data);
+    // Find the last font face loaded
+    let face = font_db
+        .faces()
+        .last()
+        .ok_or(FontThumbnailError::NoFontFound)?;
+    Ok(face.id)
+}
+
+/// Measure the text to get the size of the bounding box required
+///
+/// # Remarks
+/// The width may come back as `0.0` if the text is empty or the buffer width is
+/// too small.
+fn measure_text(
+    text: &str,
+    attrs: Attrs,
+    buffer: &mut BorrowedWithFontSystem<Buffer>,
+) -> Result<Size> {
+    buffer.set_text(text, attrs, cosmic_text::Shaping::Advanced);
+    measure_text_in_buffer(buffer)
+}
+
+/// Measure the text in the buffer to get the size of the bounding box required
+///
+/// # Remarks
+/// The width may come back as `0.0` if the text is empty or the buffer width is
+/// too small.
+fn measure_text_in_buffer(buffer: &mut BorrowedWithFontSystem<Buffer>) -> Result<Size> {
+    if buffer.size().0 < 0. || buffer.size().1 < 0. {
+        return Err(FontThumbnailError::InvalidBufferSize);
+    }
+    let mut run_width: f32 = 0.0;
+    let line_height = buffer.lines.len() as f32 * buffer.metrics().line_height;
+    let layout_runs = buffer.layout_runs();
+    for run in layout_runs {
+        run_width = run_width.max(run.line_w);
+    }
+    Ok(Size {
+        w: run_width.ceil(),
+        h: line_height,
+    })
+}
+
+impl From<Arc<Font>> for FontMetrics {
+    fn from(font: Arc<Font>) -> Self {
+        // Get the font metrics
+        let font_metrics = font.as_swash().metrics(&[]);
+        FontMetrics {
+            units_per_em: font_metrics.units_per_em as u32,
+            ascent: font_metrics.ascent,
+            descent: font_metrics.descent,
+        }
+    }
+}
+
+impl From<Arc<Font>> for FontNameInfo {
+    fn from(font: Arc<Font>) -> Self {
+        let face = font.rustybuzz();
+        let mut full_name = None;
+        let mut sample_text = None;
+        for name in face.names() {
+            match name.name_id {
+                FULL_NAME_ID => full_name = name.to_string(),
+                SAMPLE_TEXT_ID => sample_text = name.to_string(),
+                _ => {}
+            }
+        }
+        FontNameInfo {
+            full_name,
+            sample_text,
+        }
+    }
+}
+
+/// Get a Skia paint for the color, which is a solid color shader, and not using
+/// anti-aliasing
+fn get_skia_paint_for_color<'a>(color: Color) -> tiny_skia::Paint<'a> {
+    tiny_skia::Paint {
+        shader: tiny_skia::Shader::SolidColor({
+            let (r, g, b, a) = color.as_rgba_tuple();
+            tiny_skia::Color::from_rgba8(r, g, b, a)
+        }),
+        blend_mode: tiny_skia::BlendMode::Color,
+        anti_alias: false,
+        ..Default::default()
+    }
+}
+
+///  utility to generate a thumbnail from a file at path
```

Comment:
Weird comments here?

Reply:
Oops, deleted code and missed the documentation comment.